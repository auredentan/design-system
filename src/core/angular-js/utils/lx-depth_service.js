function LxDepthService() {
    const service = this;

    /////////////////////////////
    //                         //
    //    Private attributes   //
    //                         //
    /////////////////////////////

    /**
     * The initial depth.
     *
     * @type {number}
     */
    // eslint-disable-next-line no-magic-numbers
    let _depth = 1000;

    /////////////////////////////
    //                         //
    //     Public functions    //
    //                         //
    /////////////////////////////

    /**
     * Get the highest depth.
     *
     * @return {number} The highest depth.
     */
    function get() {
        return _depth;
    }

    /**
     * Increase depth.
     */
    function increase() {
        _depth++;
    }

    /////////////////////////////

    service.get = get;
    service.increase = increase;
}

/////////////////////////////

angular.module('lumx.utils.depth').service('LxDepthService', LxDepthService);

/////////////////////////////

export { LxDepthService };
