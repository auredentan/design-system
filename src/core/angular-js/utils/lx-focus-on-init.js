function lxFocusOnInitDirective() {
    function link(scope, el) {
        el.focus();
    }

    return {
        link,
    };
}

/////////////////////////////

angular.module('lumx.utils.focus-on-init').directive('lxFocusOnInit', lxFocusOnInitDirective);

/////////////////////////////

export { lxFocusOnInitDirective };
