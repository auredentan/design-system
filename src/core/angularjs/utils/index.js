/* eslint-disable import/no-unassigned-import */

import './depth_service';
import './enter-keypress_directive';
import './event-scheduler_service';
import './focus-on-init_directive';
import './focus-trap_service';
import './utils_service';
