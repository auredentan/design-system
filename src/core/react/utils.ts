import React from 'react';

import get from 'lodash/get';
import has from 'lodash/has';
import isEmpty from 'lodash/isEmpty';
import isString from 'lodash/isString';

/////////////////////////////

type Component = string | React.FC<any> | React.PureComponent<any, any> | React.Component<any, any>;

/////////////////////////////

/**
 * Check if a ReactElement is of the given type.
 *
 * @param  {React.ReactElement} el   The ReactElement we want to check the type of.
 * @param  {string|Component}   type The type we want to check if the ReactElement is of.
 * @return {boolean}            If the ReactElement is of the given type or not.
 */
function isElementOfType(el: React.ReactElement, type: Component): boolean {
    if (!isString(type) && has(type, 'displayName')) {
        type = get(type, 'displayName', get(type, '_reactInternalFiber.elementType.name', 'Component'));
    }

    if (isEmpty(type) && !isString(type)) {
        throw new Error(
            `The type you want to check is not valid. Waiting a component or a string, got ${type.toString()}`,
        );
    }

    if (type === 'text') {
        return isString(el) || el.type === 'span';
    }

    return get(el.type, 'name') === type;
}

/**
 * Check if a ReactElement is a text (i.e. either a pure text node or a <span>).
 * Simply check if the ReactElement is a string or if its type is 'span'.
 *
 * @param  {React.ReactElement} el The ReactElement to check if it's a text.
 * @return {boolean}            If the ReactElement is a text or not.
 */
function isElementText(el: React.ReactElement): boolean {
    return isElementOfType(el, 'text');
}

/////////////////////////////

export { Component, isElementOfType, isElementText };
