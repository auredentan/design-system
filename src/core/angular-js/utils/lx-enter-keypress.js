import { MODULE_NAME } from '@lumx/angularjs/lumx'

/////////////////////////////

function lxEnterKeypressDirective() {
    function link(scope, el, attrs) {
        const _ENTER_KEY_CODE = 13;

        el.bind('keydown keypress', function onKeyPress(evt) {
            if (evt.which === _ENTER_KEY_CODE) {
                scope.$apply(function evalExpression() {
                    scope.$eval(attrs.lxEnterKeypress, { $event: evt });
                });

                evt.preventDefault();
            }
        });
    }

    return {
        link,
    };
}

/////////////////////////////

angular.module(`${MODULE_NAME}.utils.enter-keypress`).directive('lxEnterKeypress', lxEnterKeypressDirective);

/////////////////////////////

export { lxEnterKeypressDirective };
