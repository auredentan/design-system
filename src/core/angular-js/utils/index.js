import './lx-depth_service';
import './lx-enter-keypress';
import './lx-event-scheduler_service';
import './lx-utils_service';
