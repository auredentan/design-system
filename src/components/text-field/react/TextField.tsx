import React, { ReactElement, RefObject, useState } from 'react';

import classNames from 'classnames';
import get from 'lodash/get';
import uuid from 'uuid/v4';

import { Icon, Size, Theme } from 'LumX';
import { CSS_PREFIX } from 'LumX/core/constants';
import { COMPONENT_PREFIX } from 'LumX/core/react/constants';
import { IGenericProps, getRootClassName } from 'LumX/core/react/utils';
import { handleBasicClasses } from 'LumX/core/utils';
import { mdiAlertCircle, mdiCheckCircle } from 'LumX/icons';

/////////////////////////////

enum TextFieldType {
    input = 'input',
    textarea = 'textarea',
}

/**
 * Defines the props of the component.
 */
interface ITextFieldProps extends IGenericProps {
    /** Whether the text field is displayed with error style or not. */
    hasError?: boolean;

    /** Text field helper message. */
    helper?: string;

    /** Text field icon (SVG path). */
    icon?: string;

    /** Id that will be passed to input element. An id is generated (uuid) if no id is provided. */
    id?: string;

    /** Whether the text field is disabled or not. */
    isDisabled?: boolean;

    /** Whether the text field is displayed with valid style or not. */
    isValid?: boolean;

    /** Text field label displayed in a label tag. */
    label?: string;

    /** Text field placeholder message. */
    placeholder?: string;

    /** Theme. */
    theme?: Theme;

    /** A ref that will be passed to the wrapper element. */
    textFieldRef?: RefObject<HTMLDivElement>;

    /** A ref that will be passed to the input or text area element. */
    inputRef?: RefObject<HTMLInputElement> | RefObject<HTMLTextAreaElement>;

    /** Whether custom colors are applied to this component. */
    useCustomColors?: boolean;

    /** Text field value. */
    value: string;

    /** Text field type (input or textarea). */
    type?: TextFieldType;

    /** Text field value change handler. */
    onChange(value: string): void;

    /** Text field focus change handler. */
    onFocus?(event: React.FocusEvent): void;

    /** Text field blur change handler. */
    onBlur?(event: React.FocusEvent): void;
}
type TextFieldProps = ITextFieldProps;

/////////////////////////////

/////////////////////////////
//                         //
//    Public attributes    //
//                         //
/////////////////////////////

/**
 * The display name of the component.
 */
const COMPONENT_NAME = `${COMPONENT_PREFIX}TextField`;

/**
 * The default class name and classes prefix for this component.
 */
const CLASSNAME: string = getRootClassName(COMPONENT_NAME);

/**
 * The default value of props.
 */
const DEFAULT_PROPS: Partial<TextFieldProps> = {
    hasError: false,
    isDisabled: false,
    isValid: false,
    theme: Theme.light,
    type: TextFieldType.input,
};

/////////////////////////////

interface IInputNativeProps {
    id?: string;
    isDisabled?: boolean;
    placeholder?: string;
    type?: TextFieldType;
    value: string;
    inputRef?: RefObject<HTMLInputElement> | RefObject<HTMLTextAreaElement>;
    setFocus(focus: boolean): void;
    onChange(value: string): void;
    onFocus?(value: React.FocusEvent): void;
    onBlur?(value: React.FocusEvent): void;
}

const renderInputNative = (props: IInputNativeProps): ReactElement => {
    const {
        id,
        isDisabled,
        placeholder,
        type,
        value,
        setFocus,
        onChange,
        onFocus,
        onBlur,
        inputRef,
        ...forwardedProps
    } = props;

    const onTextFieldFocus = (event: React.FocusEvent): void => {
        if (onFocus) {
            onFocus(event);
        }

        return setFocus(true);
    };

    const onTextFieldBlur = (event: React.FocusEvent): void => {
        if (onBlur) {
            onBlur(event);
        }

        return setFocus(false);
    };

    const handleChange = (event: React.ChangeEvent): void => onChange(get(event, 'target.value'));

    if (type === TextFieldType.textarea) {
        return (
            <textarea
                id={id}
                disabled={isDisabled}
                placeholder={placeholder}
                value={value}
                onFocus={onTextFieldFocus}
                onBlur={onTextFieldBlur}
                onChange={handleChange}
                ref={inputRef as RefObject<HTMLTextAreaElement>}
                {...forwardedProps}
            />
        );
    }
    return (
        <input
            id={id}
            disabled={isDisabled}
            type="text"
            placeholder={placeholder}
            value={value}
            onFocus={onTextFieldFocus}
            onBlur={onTextFieldBlur}
            onChange={handleChange}
            ref={inputRef as RefObject<HTMLInputElement>}
            {...forwardedProps}
        />
    );
};

/**
 * Text field.
 *
 * @param  props Text field props.
 * @return The component.
 */
const TextField: React.FC<TextFieldProps> = (props: TextFieldProps): ReactElement => {
    const {
        className = '',
        hasError,
        helper,
        icon,
        id = uuid(),
        isDisabled,
        isValid,
        label,
        onChange,
        onFocus,
        onBlur,
        placeholder,
        textFieldRef,
        inputRef,
        theme = DEFAULT_PROPS.theme,
        type = DEFAULT_PROPS.type,
        useCustomColors,
        value,
        ...forwardedProps
    } = props;
    const [isFocus, setFocus] = useState(false);

    return (
        <div
            className={classNames(
                className,
                handleBasicClasses({
                    hasError: !isValid && hasError,
                    hasIcon: Boolean(icon),
                    hasInput: type === TextFieldType.input,
                    hasLabel: Boolean(label),
                    hasPlaceholder: Boolean(placeholder),
                    hasTextarea: type === TextFieldType.textarea,
                    hasValue: Boolean(value),
                    isDisabled,
                    isFocus,
                    isValid,
                    prefix: CLASSNAME,
                    theme,
                }),
                { [`${CSS_PREFIX}-custom-colors`]: useCustomColors },
            )}
            ref={textFieldRef}
        >
            {label && (
                <label htmlFor={id} className={`${CLASSNAME}__label`}>
                    {label}
                </label>
            )}

            {helper && <span className={`${CLASSNAME}__helper`}>{helper}</span>}

            <div className={`${CLASSNAME}__input-wrapper`}>
                {icon && (
                    <Icon
                        className={`${CLASSNAME}__input-icon`}
                        color={theme === Theme.dark ? 'light' : undefined}
                        icon={icon}
                        size={Size.xs}
                    />
                )}

                <div className={`${CLASSNAME}__input-native`}>
                    {renderInputNative({
                        id,
                        inputRef,
                        isDisabled,
                        onBlur,
                        onChange,
                        onFocus,
                        placeholder,
                        setFocus,
                        type,
                        value,
                        ...forwardedProps,
                    })}
                </div>

                {(isValid || hasError) && (
                    <Icon
                        className={`${CLASSNAME}__input-validity`}
                        color={theme === Theme.dark ? 'light' : undefined}
                        icon={isValid ? mdiCheckCircle : mdiAlertCircle}
                        size={Size.xs}
                    />
                )}
            </div>
        </div>
    );
};
TextField.displayName = COMPONENT_NAME;

/////////////////////////////

export { CLASSNAME, DEFAULT_PROPS, TextField, TextFieldType, TextFieldProps };
