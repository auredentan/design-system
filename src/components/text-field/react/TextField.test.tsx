import React, { ReactElement } from 'react';

import { mount, shallow } from 'enzyme';
import { build } from 'test-data-bot';

import { ICommonSetup, Wrapper, commonTestsSuite } from 'LumX/core/testing/utils.test';
import { getBasicClass } from 'LumX/core/utils';

import { CLASSNAME, TextField, TextFieldProps } from './TextField';

/////////////////////////////

/**
 * Define the overriding properties waited by the `setup` function.
 */
type ISetupProps = Partial<TextFieldProps>;

/**
 * Defines what the `setup` function will return.
 */
interface ISetup extends ICommonSetup {
    props: ISetupProps;

    /**
     * The <div> element that wraps checkbox and children elements.
     */
    wrapper: Wrapper;
}

/////////////////////////////

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 *
 * @param props  The props to use to override the default props of the component.
 * @param     [shallowRendering=true] Indicates if we want to do a shallow or a full rendering.
 * @return      An object with the props, the component wrapper and some shortcut to some element inside of the
 *                       component.
 */
const setup = ({ ...propsOverrides }: ISetupProps = {}, shallowRendering: boolean = true): ISetup => {
    const props: TextFieldProps = {
        onChange: jest.fn(),
        value: 'Test value',
        ...propsOverrides,
    };

    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;

    // Noinspection RequiredAttributes.
    const wrapper: Wrapper = renderer(<TextField {...props} />);

    return {
        props,
        wrapper,
    };
};

describe(`<${TextField.displayName}>`, (): void => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', (): void => {
        it('should render correctly', (): void => {
            const { wrapper } = setup({ id: 'fixedId' });
            expect(wrapper).toMatchSnapshot();

            expect(wrapper).toExist();

            expect(wrapper).toHaveClassName(CLASSNAME);
            expect(wrapper).not.toHaveClassName('lumx-text-field--is-valid');
            expect(wrapper).not.toHaveClassName('lumx-text-field--has-error');
            expect(wrapper).not.toHaveClassName('lumx-text-field--has-label');
            expect(wrapper).not.toHaveClassName('lumx-text-field--is-disabled');
            expect(wrapper).not.toHaveClassName('lumx-text-field--has-placeholder');
            expect(wrapper).not.toHaveClassName('lumx-text-field--is-focus');
            expect(wrapper).not.toHaveClassName('lumx-text-field--has-icon');

            expect(wrapper).toHaveClassName('lumx-text-field--theme-light');
        });
    });

    /////////////////////////////

    // 2. Test defaultProps value and important props custom values.
    describe('Props', (): void => {
        it('should add all class names (except has-error)', (): void => {
            const modifiedPropsBuilder: () => ISetupProps = build('props').fields!({
                icon: 'icon',
                isDisabled: true,
                isValid: true,
                label: 'test',
                placeholder: 'test',
            });

            const modifiedProps: ISetupProps = modifiedPropsBuilder();

            const { wrapper } = setup({ ...modifiedProps });

            Object.keys(modifiedProps).forEach(
                (prop: string): void => {
                    const propType =
                        prop === 'icon' || prop === 'label' || prop === 'placeholder'
                            ? `has${prop.charAt(0).toUpperCase() + prop.slice(1)}`
                            : prop;
                    const propValue =
                        prop === 'icon' || prop === 'label' || prop === 'placeholder' ? true : modifiedProps[prop];
                    expect(wrapper).toHaveClassName(
                        getBasicClass({ prefix: CLASSNAME, type: propType, value: propValue }),
                    );
                },
            );
        });

        it('should add "has-error" class name', (): void => {
            const modifiedPropsBuilder: () => ISetupProps = build('props').fields!({
                hasError: true,
            });

            const modifiedProps: ISetupProps = modifiedPropsBuilder();

            const { wrapper } = setup({ ...modifiedProps });

            Object.keys(modifiedProps).forEach(
                (prop: string): void => {
                    expect(wrapper).toHaveClassName(
                        getBasicClass({ prefix: CLASSNAME, type: prop, value: modifiedProps[prop] }),
                    );
                },
            );
        });
    });

    /////////////////////////////

    // 3. Test events.
    describe('Events', (): void => {
        const onChange: jest.Mock = jest.fn();

        beforeEach(
            (): void => {
                onChange.mockClear();
            },
        );

        it('should trigger `onChange` when text field is changed', (): void => {
            const onChangeMock = jest.fn();
            const event = {
                target: { value: 'my value' },
            };
            const component = shallow(<TextField value="" onChange={onChangeMock} />);
            component.find('input').simulate('change', event);
            expect(onChangeMock).toBeCalledWith('my value');
        });
    });

    /////////////////////////////

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', (): void => {
        // Nothing to do here.
    });

    /////////////////////////////

    // 5. Test state.
    describe('State', (): void => {
        // Nothing to do here.
    });

    /////////////////////////////

    // Common tests suite.
    commonTestsSuite(setup, {}, { className: CLASSNAME });
});
