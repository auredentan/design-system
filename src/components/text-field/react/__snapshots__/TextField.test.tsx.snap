// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField> Snapshots and structure should render correctly 1`] = `
<div
  className="lumx-text-field lumx-text-field--has-value lumx-text-field--theme-light"
>
  <div
    className="lumx-text-field__input-wrapper"
  >
    <div
      className="lumx-text-field__input-native"
    >
      <input
        id="fixedId"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="text"
        value="Test value"
      />
    </div>
  </div>
</div>
`;
